streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/HighAvailabilityTaskAssignor.java

```diff
@@ -0,0 +1,544 @@
+
+        final boolean followupRebalanceRequired;
+        if (shouldUsePreviousAssignment()) {
+            assignPreviousTasksToClientStates();
+            followupRebalanceRequired = false;
+        } else {
+            assignActiveTasksToClients(statefulActiveTaskAssignment);
+            assignStandbyTasksToClients(warmupTaskAssignment);
+            assignStandbyTasksToClients(standbyTaskAssignment);
+            assignActiveTasksToClients(statelessActiveTaskAssignment);
+
+            followupRebalanceRequired = warmupTaskAssignment.entrySet().stream().anyMatch(tasks -> !tasks.getValue().isEmpty());
+        }
+        return followupRebalanceRequired;
+    }
+
+    /**
+     * Returns a list of the movements of tasks from statefulActiveTaskAssignment to balancedStatefulActiveTaskAssignment
+     * @param statefulActiveTaskAssignment the initial assignment, with source clients
+     * @param balancedStatefulActiveTaskAssignment the final assignment, with destination clients
+     */
+    static <ID> Queue<Movement<ID>> getMovements(final Map<ID, List<TaskId>> statefulActiveTaskAssignment,
+                                                 final Map<ID, List<TaskId>> balancedStatefulActiveTaskAssignment) {
+        if (statefulActiveTaskAssignment.size() != balancedStatefulActiveTaskAssignment.size()) {
+            throw new IllegalStateException("Tried to compute movements but assignments differ in size.");
+        }
+
+        final Map<TaskId, ID> taskToDestinationClient = new HashMap<>();
+        for (final Map.Entry<ID, List<TaskId>> clientEntry : balancedStatefulActiveTaskAssignment.entrySet()) {
+            final ID destination = clientEntry.getKey();
+            for (final TaskId task : clientEntry.getValue()) {
+                taskToDestinationClient.put(task, destination);
+            }
+        }
+
+        final Queue<Movement<ID>> movements = new LinkedList<>();
+        for (final Map.Entry<ID, List<TaskId>> sourceClientEntry : statefulActiveTaskAssignment.entrySet()) {
+            final ID source = sourceClientEntry.getKey();
+
+            for (final TaskId task : sourceClientEntry.getValue()) {
+                final ID destination = taskToDestinationClient.get(task);
+                if (destination == null) {
+                    log.error("Task {} is assigned to client {} in initial assignment but has no owner in the final " +
+                                  "balanced assignment.", task, source);
+                    throw new IllegalStateException("Found task in initial assignment that was not assigned in the final.");
+                } else if (source != destination) {
+                    movements.add(new Movement<>(task, source, destination));
+                }
+            }
+        }
+        return movements;
+    }
+
+    /**
+     * @return true iff all active tasks with caught-up client are assigned to one of them, and all tasks are assigned
+     */
+    boolean previousAssignmentIsValid() {
```

Comment:
Q: There are a lot of methods that are package-private for testing. I would avoid to do that too much. If we feel we need to test sub components of a class then we might want to factor this components out to their own classes that we can test separately. For example, the validation of the previous assignment can be encapsulated into its own class. Does not need to be in this PR, though. WDYT?

Reply:
For my part, I think that several of the static inner classes of this assignor should be pulled out into top-level classes, which would have the side effect of pulling several of these tests that are unrelated to the HATA's contract itself. But I don't think it's high value right now, as I also want to refactor this module a bit before we finalize the feature.

As far as having visible-for-testing methods, I agree it's generally a failure of encapsulation, but in this particular case, we don't rely on the HATA's built-in public interface, but always access it via the `TaskAssignor` interface in the production code path, which makes package-visible methods in HATA much more palatable. It's also mitigating the risk that this method doesn't mutate the assignor, so I don't find this particular case too concerning.

Altogether, if we had a package-visible static method on another object to do this, or a package-visible instance method on this object, it seems to be equally visible, so I guess it's no practical difference in this particular case. If you really want to encapsulate the method, the solution for both approaches is to put the HATA in its own package: `o.a.k.s.p.i.assignment.ha.*`, but it feels like diminishing returns.

I _do_ look forward to dropping support for Java 8 so that we can declare module public interfaces and not have to contend with compromises like this.

Reply:
I agree with you that in this specific case, it might not be that concerning and that we can leave refactoring as a follow-up. However, I see the practice visible for testing a lot across the code base and I would like to keep at the absolute minimum in new code.

Reply:
Oh, yeah, I'm 100% on board with you there. `// visible for testing` always reads to me like `// setting a trap for the future:`.